nearcache_map.go

```diff
@@ -0,0 +1,265 @@
+/*
+ * Copyright (c) 2008-2021, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License")
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package hazelcast
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/serialization"
+	"github.com/hazelcast/hazelcast-go-client/nearcache"
+)
+
+type nearCacheMap struct {
+	nc             *nearCache
+	ncc            *nearcache.Config
+	toNearCacheKey func(key interface{}) (interface{}, error)
+	ss             *serialization.Service
+}
+
+func newNearCacheMap(nc *nearCache, ncc *nearcache.Config, ss *serialization.Service) (nearCacheMap, error) {
+	ncm := nearCacheMap{
+		nc:  nc,
+		ncc: ncc,
+		ss:  ss,
+	}
+	// the only valid local policy on the client side is invalidate.
+	ncm.registerInvalidationListener()
+	if ncc.Preloader.Enabled {
+		if err := ncm.preload(); err != nil {
+			return nearCacheMap{}, fmt.Errorf("preloading near cache: %w", err)
+		}
+	}
+	// toNearCacheKey returns the raw key if SerializeKeys is not true.
+	if ncc.SerializeKeys {
+		ncm.toNearCacheKey = func(key interface{}) (interface{}, error) {
+			data, err := ss.ToData(key)
+			if err != nil {
+				return nil, err
+			}
+			// byte slices cannot be map keys
+			return data, nil
+		}
+	} else {
+		ncm.toNearCacheKey = func(key interface{}) (interface{}, error) {
+			return key, nil
+		}
+	}
+	return ncm, nil
+}
+
+func (ncm *nearCacheMap) registerInvalidationListener() {
+	fmt.Println("IMPLEMENT ME: registerInvalidationListener")
+}
+
+func (ncm *nearCacheMap) preload() error {
+	panic("implement me!")
+}
+
+func (ncm *nearCacheMap) ContainsKey(ctx context.Context, key interface{}, m *Map) (found bool, err error) {
+	key, err = ncm.toNearCacheKey(key)
+	if err != nil {
+		return false, err
+	}
+	cached, ok, err := ncm.getCachedValue(key, false)
+	if err != nil {
+		return false, err
+	}
+	if ok {
+		return cached != nil, nil
+	}
+	keyData, err := m.validateAndSerialize(key)
+	if err != nil {
+		return false, err
+	}
+	return m.containsKeyFromRemote(ctx, keyData)
```

Comment:
Does it makes sense to add a pointer to the map proxy to nearCacheMap and introduce a new method called something like `containsKeyInternal` and call it here instead of copying the code to each method. It will be hard to update. 

So,

ContainsKey in proxy_map.go:

```go
// ContainsKey returns true if the map contains an entry with the given key.
func (m *Map) ContainsKey(ctx context.Context, key interface{}) (bool, error) {
	if m.hasNearCache {
		return m.ncm.ContainsKey(ctx, key, m)
	}
	return m.containsInternal(ctx, key)
}

// ContainsKey returns true if the map contains an entry with the given key.
func (m *Map) containsInternal(ctx context.Context, key interface{}) (bool, error) {
        keyData, err := m.validateAndSerialize(key)
	if err != nil {
		return false, err
	}
	return m.containsKeyFromRemote(ctx, keyData)
}
``` 

ContainsKey in nearcache_map.go:

```go
func (ncm *nearCacheMap) ContainsKey(ctx context.Context, key interface{}, m *Map) (found bool, err error) {
	key, err = ncm.toNearCacheKey(key)
	if err != nil {
		return false, err
	}
	cached, ok, err := ncm.getCachedValue(key, false)
	if err != nil {
		return false, err
	}
	if ok {
		return cached != nil, nil
	}
	return ncm.m.containsKeyInternal(ctx, key)
}
```

Similar for other methods.